src/rt.rs

```diff
@@ -0,0 +1,51 @@
+// Copyright 2013-2018, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use std::cell::Cell;
+use std::sync::atomic::{AtomicBool, ATOMIC_BOOL_INIT, Ordering};
+
+thread_local! {
+    static IS_MAIN_THREAD: Cell<bool> = Cell::new(false)
+}
+
+static INITIALIZED: AtomicBool = ATOMIC_BOOL_INIT;
+
+/// Asserts that this is the main thread and `gtk::init` has been called.
+macro_rules! assert_initialized_main_thread {
+    () => (
+        if !::rt::is_initialized_main_thread() {
+            if ::rt::is_initialized() {
+                panic!("GTK may only be used from the main thread.");
+            }
+            else {
+                panic!("GTK has not been initialized. Call `gtk::init` first.");
+            }
+        }
+    )
+}
+
+macro_rules! callback_guard {
+    () => (
+        let _guard = ::glib::CallbackGuard::new();
+    )
+}
+
+/// No-op.
+macro_rules! skip_assert_initialized {
+    () => ()
+}
+
+/// Returns `true` if Sourceview has been initialized.
+#[inline]
+pub fn is_initialized() -> bool {
+    skip_assert_initialized!();
+    INITIALIZED.load(Ordering::Acquire)
+}
+
+/// Returns `true` if Sourceview has been initialized and this is the main thread.
+#[inline]
+pub fn is_initialized_main_thread() -> bool {
+    skip_assert_initialized!();
+    IS_MAIN_THREAD.with(|c| c.get())
```

Comment:
Both booleans are never actually set to `true` or am I missing something?

You basically need to ensure that `IS_MAIN_THREAD` is only `true` on the thread on which GTK was initialized, and `INITIALIZED` must be `true` once GTK is initialized.

Reply:
Oups indeed! Fixing it.